haskell-best-practices.md

```diff
@@ -184,6 +184,172 @@ instance Monoid Mult where
   mappend = (*)
 ```
 
+## Phantom types
+
+The `Tagged` example above makes use of a phantom type variable. The definition of `Tagged` is:
+```haskell
+newtype Tagged t a = Tagged { untag :: a }
+```
+
+Note how the type variable `t` does not appear on the right-hand-side of the equals-sign. It's only used to add type-level information to a type - it's never used at the value level. We call this kind of type variable a phantom type variable, or occasionally, just a phantom type.
+
+A more useful example is hashing passwords. We might want to represent passwords as either raw text or hashed text and have operations that only work on one kind:
+
+```haskell
+module Password (Password, PasswordState, rawPassword, hashPassword, comparePassword) where
+
+import Data.Hash
+
+
+data PasswordState = Raw | Hashed
+
+newtype Password (s :: PasswordState) = Password Text
+
+rawPassword :: Text -> Password 'Raw
+rawPassword = Password
+
+-- Note: this hash is may be less secure than advertised
+secureHash :: Text -> Text
+secureHash t = reverse $ t ++ t
+
+hashPassword :: Password 'Raw -> Password 'Hashed
+hashPassword (Password r) = Password (secureHash r)
+
+comparePassword :: Password 'Raw -> Password 'Hashed -> Bool
+comparePassword (Password r) (Password h) = secureHash r == h
+```
+
+We're using a promoted datatype (`PasswordState`) to encode whether a `Password` is plaintext or hashed. Note that we do not expose the constructor for `Password`, and instead expose a smart constructor `rawPassword` and a hashing function `hashPassword`. If we expose the constructor, a malicious (or forgetful) user could construct a value of `Password 'Hashed` that contains a plaintext password.
+
+## GADTs
+
+Closely related to phantom types are Generalized Algebraic Data Types or GADTs. Suppose we had the following representation of a small programming language with integers, bools, addition, and conditions:
+```haskell
+data Expr
+  | I Int
+  | B Bool
+  | Add Expr Expr
+  | LessThan Expr Expr
+  | Cond Expr Expr Expr
+    deriving (Eq, Show)
+```
+
+We can construct values like `Add (I 1) (I 3)` to represent `1 + 3`, but there's nothing preventing us from constructing values like `Add (I 1) (B True)`. Writing an evaluation function for this data type is frustrating:
+
+```haskell
+data Value
+  = VI Int
+  | VB Bool
+
+eval :: Expr -> Maybe Value
+eval (I i) = return $ VI i
+eval (B b) = return $ VB b
+eval (Add x y) =
+  VI x' <- eval x
+  VI y' <- eval y
+  return $ VI $ x' + y'
+eval (LessThan x y) =
+  VI x' <- eval x
+  VI y' <- eval y
+  return $ VB $ x' < y'
+eval (Cond c t f) =
+  VB c' <- eval c
+  if c'
+    then eval t
+    else eval f
+```
+
+We're relying on the `Monad` instance for for `Maybe` to call `fail` on pattern match failures when we pass something like `Add (I 1) (B True)`. We also have to make an extra datatype to represent values.
```

Comment:
Also think focus should be on pointing out that this is a run-time exception vs something we can detect at compilation